### ISSUE TITLE
Access logging doesn't go to file when behind sslh

### ISSUE THREAD

**Issue by HandyHat** [caddyserver/caddy#5656]
When making a request via [sslh](https://github.com/yrutschle/sslh), rather than logging to a file specified in the Caddyfile, the request is logged to the `default` logger (i.e. console by default).

This can be minimally reproduced using the following Caddyfile:
```Caddyfile
https://localhost:4443 {
	log {
		output file /home/pi/caddy.log
	}
	file_server
}
```
and sslh config:

```
sslh --listen 0.0.0.0:443 --tls 127.0.0.1:4443 --openvpn 127.0.0.1:8443
```

Navigating to https://localhost:4443 correctly logs the request to `home/pi/caddy.log`:
```
2023/07/27 11:11:09.237	info	http.log.access.log0	handled request	{"request": {"remote_ip": "::1", "remote_port": "59742", "client_ip": "::1", "proto": "HTTP/2.0", "method": "GET", "host": "localhost:4443", "uri": "/", "headers": {"Accept": ["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"], "Sec-Fetch-Mode": ["navigate"], "Accept-Encoding": ["gzip, deflate, br"], "Sec-Ch-Ua": ["\"Chromium\";v=\"113\", \"Not-A.Brand\";v=\"24\""], "Sec-Ch-Ua-Mobile": ["?0"], "Upgrade-Insecure-Requests": ["1"], "User-Agent": ["Mozilla/5.0 (X11; CrOS aarch64 13597.84.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.5672.95 Safari/537.36"], "Sec-Fetch-Dest": ["document"], "Accept-Language": ["en-GB,en-US;q=0.9,en;q=0.8"], "Cache-Control": ["max-age=0"], "Sec-Ch-Ua-Platform": ["\"Linux\""], "Sec-Fetch-Site": ["none"], "Sec-Fetch-User": ["?1"]}, "tls": {"resumed": false, "version": 772, "cipher_suite": 4867, "proto": "h2", "server_name": "localhost"}}, "bytes_read": 0, "user_id": "", "duration": 0.015666546, "size": 13, "status": 200, "resp_headers": {"Alt-Svc": ["h3=\":4443\"; ma=2592000"], "Etag": ["\"ryfjbmd\""], "Content-Type": ["text/html; charset=utf-8"], "Last-Modified": ["Thu, 27 Jul 2023 00:42:58 GMT"], "Accept-Ranges": ["bytes"], "Content-Length": ["13"], "Server": ["Caddy"]}}
```

Navigating to https://localhost:443, however, incorrect logs the request to the console rather than the file even though sslh proxies the request to 127.0.0.1:4443:

```
2023/07/27 11:11:31.950 INFO    http.log.access handled request {"request": {"remote_ip": "127.0.0.1", "remote_port": "34226", "client_ip": "127.0.0.1", "proto": "HTTP/2.0", "method": "GET", "host": "localhost", "uri": "/", "headers": {"Sec-Ch-Ua-Mobile": ["?0"], "Accept-Language": ["en-GB,en-US;q=0.9,en;q=0.8"], "Sec-Fetch-User": ["?1"], "Sec-Ch-Ua": ["\"Chromium\";v=\"113\", \"Not-A.Brand\";v=\"24\""], "Sec-Ch-Ua-Platform": ["\"Linux\""], "Upgrade-Insecure-Requests": ["1"], "User-Agent": ["Mozilla/5.0 (X11; CrOS aarch64 13597.84.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.5672.95 Safari/537.36"], "Accept": ["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"], "Sec-Fetch-Site": ["none"], "Sec-Fetch-Mode": ["navigate"], "Sec-Fetch-Dest": ["document"], "Accept-Encoding": ["gzip, deflate, br"]}, "tls": {"resumed": false, "version": 772, "cipher_suite": 4867, "proto": "h2", "server_name": "localhost"}}, "bytes_read": 0, "user_id": "", "duration": 0.000256889, "size": 13, "status": 200, "resp_headers": {"Content-Length": ["13"], "Server": ["Caddy"], "Alt-Svc": ["h3=\":4443\"; ma=2592000"], "Etag": ["\"ryfjbmd\""], "Content-Type": ["text/html; charset=utf-8"], "Last-Modified": ["Thu, 27 Jul 2023 00:42:58 GMT"], "Accept-Ranges": ["bytes"]}}
```

This occurs on caddy `v2.6.4` and `v2.7.0-beta.2 h1:jaS1odoRuDR2W8igaKgVGvVjhTNt8xfoz3YPC4bcenA=`

**Comment by mholt** [comment]
This is probably because the Host header `localhost` (implied `:443`, I assume)  doesn't match `localhost:4443` so the logging configuration isn't applied.

If you submit the same request but force the Host header to be `localhost:4443` (or have sslh rewrite it, if possible -- a proper HTTP proxy should have that ability) then it will probably work as expected.

(Closing, since I don't think this is a bug in Caddy. It seems like a misconfiguration of a downstream application. But we can reopen if needed.)

**Comment by HandyHat** [comment]
Thanks for clarifying @mholt!
Would it be possible to make this clearer in the documentation for logging? I spent a long time trying to figure why this wasn't working and I don't think I missed anything while reading through the docs (though I may have!)

**Comment by francislavoie** [comment]
I think it might be clearer now actually, just by the virtue of adding a new `hostnames` config option which itself has some documentation https://caddyserver.com/docs/caddyfile/directives/log#syntax. Though I guess we could mention specifically that port numbers are included in that.

**Comment by mholt** [comment]
Sure, we could probably clarify this (typically, the "Host" does include a port, if present, so I figured that was pretty standard).

**Comment by HandyHat** [comment]
Ah, this is perfect - thank you!

**Comment by clonejo** [comment]
I feel it is extremely confusing for logging decisions to be different from actual request handling. Requests get handled regardless of the `Host` header, using the site blocks. But for logging suddenly the `Host` header matters.

For the current behavior, it would be more natural if `log` directives had to be specified outside the site blocks.

Notably, this magic behavior is not visible in the `caddy adapt` output.

**Comment by mholt** [comment]
@clonejo I'm not quite sure I follow. Mind elaborating?

> Requests get handled regardless of the port in the Host header, using the site blocks.

I don't see how that's the case; for example, a site block defined as `localhost:1234` will only handle requests with a header of `Host: localhost:1234`.

**Comment by clonejo** [comment]
> @clonejo I'm not quite sure I follow. Mind elaborating?
> 
> > Requests get handled regardless of the port in the Host header, using the site blocks.
> 
> I don't see how that's the case; for example, a site block defined as `localhost:1234` will only handle requests with a header of `Host: localhost:1234`.

Nope, in Caddy 2.7.5, a server block `localhost:1234` will also handle connections to port 1234 with the header `Host: localhost` (without port). But confusingly, the logging will be handled outside the server block in that case, using the default logger.

**Comment by francislavoie** [comment]
If I could start over (or when we make Caddy v3), I would make the `log` directive be called `access_log` instead, to clarify that that's what it configures. The `log` global option configures global/runtime logging (and it can also redirect where access logs are written if changing the default logger).

We didn't have a solid idea of how we wanted logging to look when v2 was initially being worked on, so a lot of the config surface was tacked-on after the fact. So I agree it's somewhat messy.

> Nope, in Caddy 2.7.5, a server block `localhost:1234` will also handle connections to port 1234 with the header `Host: localhost` (without port). But confusingly, the logging will be handled outside the server block in that case, using the default logger.

Yeah so the way it works is that `localhost:1234` in the Caddyfile is sugar for saying "make a server that listens on all interfaces on port `1234` (if no such server already exists), and add a route that matches the `Host` header for value `localhost`".

The `host` matcher ignores the port on the request because it's not relevant to matching because the TCP connection itself is what determines the port that was used.

Basically you're asking about an edgecase of the HTTP spec (being able to send a Host header which disagrees with the TCP connection) which is something browsers or normal clients never do, it only happens with a carefully crafted command/client to override the default behaviour.

**Comment by mholt** [comment]
The Host header is required to have the port if it is not the default port.
